File: week5/DiamondProblem.cpp
#include<iostream>
using namespace std;

class Class1
{
    public:
        void class1()
        {
            cout<<"Iam Class1\n";
        }
};

class Class2: public Class1
{
    public:
        void class2()
        {
            cout<<"Iam Class2\n";
        }
};

class Class3: public Class1
{
    public:
        void class3()
        {
            cout<<"Iam Class3\n";
        }
};

class Class4: public Class3,public Class2
{
    public:
        void class4()
        {
            cout<<"Iam Class4\n";
        }
};

int main()
{
    Class4 obj1;
    //obj1.class1();
    //class1() method becomes ambiguous here
    obj1.class2();
    obj1.class3();
    obj1.class4();
    return 0;
}

File: week5/DiamondProblem.java
class Class1
{
    public
        void class1()
        {
            System.out.print("Iam Class1\n");
        }
};

class Class2 extends Class1
{
    public
        void class2()
        {
            System.out.print("Iam Class2\n");
        }
};

class Class3 extends Class1
{
    public
        void class3()
        {
            System.out.print("Iam Class3\n");
        }
};

/*class Class4 extends Class2,Class3
{
    public
        void class4()
        {
            System.out.print("Iam Class4\n");
        }
};*/

public class DiamondProblem
{
    
    public static void main(String[] args)
    {
        System.out.print("Diamond Problem\n");
        Class2 obj=new Class2();
        Class3 obj1=new Class3();
        obj.class1();
        obj.class2();
        obj1.class1();
        obj1.class3();
    }
}
